File: native/ios/PaymentAdapter.swift
import Foundation

struct NativePaymentRequestModel {
    let provider: String
    let orderId: String
    let amount: Double
    let currency: String
    let raw: [String: Any]
}

struct NativePaymentResponseModel {
    let status: String
    let transactionId: String?
    let providerReference: String?
    let message: String?

    func dictionary() -> [String: Any] {
        var out: [String: Any] = ["status": status]
        if let transactionId { out["transactionId"] = transactionId }
        if let providerReference { out["providerReference"] = providerReference }
        if let message { out["message"] = message }
        return out
    }
}

@MainActor
protocol PaymentAdapter {
    var providerId: String { get }
    func start(_ request: NativePaymentRequestModel) async -> NativePaymentResponseModel
}

@MainActor
final class UnconfiguredPaymentAdapter: PaymentAdapter {
    let providerId: String
    init(providerId: String) { self.providerId = providerId }

    func start(_ request: NativePaymentRequestModel) async -> NativePaymentResponseModel {
        NativePaymentResponseModel(status: "failed", transactionId: nil, providerReference: providerId, message: "provider_sdk_not_installed")
    }
}

File: native/android/PaymentAdapter.kt
package com.subil.customer

import org.json.JSONObject

data class NativePaymentRequestModel(
    val provider: String,
    val orderId: String,
    val amount: Double,
    val currency: String,
    val raw: JSONObject
)

data class NativePaymentResponseModel(
    val status: String,
    val transactionId: String? = null,
    val providerReference: String? = null,
    val message: String? = null
) {
    fun toJson(): JSONObject = JSONObject().put("status", status).also { out ->
        transactionId?.let { out.put("transactionId", it) }
        providerReference?.let { out.put("providerReference", it) }
        message?.let { out.put("message", it) }
    }
}

interface PaymentAdapter {
    val providerId: String
    fun start(request: NativePaymentRequestModel): NativePaymentResponseModel
}

class UnconfiguredPaymentAdapter(override val providerId: String) : PaymentAdapter {
    override fun start(request: NativePaymentRequestModel) = NativePaymentResponseModel(
        status = "failed",
        providerReference = providerId,
        message = "provider_sdk_not_installed"
    )
}
